rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -5098,3 +5098,87 @@ extern "C" rmw_ret_t rmw_qos_profile_check_compatible(
   return rmw_dds_common::qos_profile_check_compatible(
     publisher_profile, subscription_profile, compatibility, reason, reason_size);
 }
+
+extern "C" rmw_ret_t rmw_client_get_actual_qos(
+  const rmw_client_t * client,
+  rmw_qos_profile_t * qos)
+{
+  RMW_CHECK_ARGUMENT_FOR_NULL(client, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(qos, RMW_RET_INVALID_ARGUMENT);
+
+  auto info = static_cast<CddsClient *>(client->data);
+
+  rmw_qos_profile_t pub_qos = rmw_qos_profile_default;
+  rmw_qos_profile_t sub_qos = rmw_qos_profile_default;
+
+  // Check if the QoS of the client's pub/sub match
+  if (get_readwrite_qos(info->client.pub->enth, &pub_qos)) {
+    if (get_readwrite_qos(info->client.sub->enth, &sub_qos)) {
+      if (pub_qos.history == sub_qos.history &&
+        pub_qos.depth == sub_qos.depth &&
+        pub_qos.reliability == sub_qos.reliability &&
+        pub_qos.durability == sub_qos.durability &&
+        pub_qos.liveliness == sub_qos.liveliness &&
+        pub_qos.deadline.sec == sub_qos.deadline.sec &&
+        pub_qos.deadline.nsec == sub_qos.deadline.nsec &&
+        pub_qos.lifespan.sec == sub_qos.lifespan.sec &&
+        pub_qos.lifespan.nsec == sub_qos.lifespan.nsec &&
+        pub_qos.liveliness_lease_duration.sec == sub_qos.liveliness_lease_duration.sec &&
+        pub_qos.liveliness_lease_duration.nsec == sub_qos.liveliness_lease_duration.nsec)
+      {
+        // The client has a single QoS, we can set it either as the qos
+        // of the publisher or subscriber, as they match.
+        get_readwrite_qos(info->client.pub->enth, qos);
```

Comment:
Nitpick: you already have a copy of the QoS settings in `pub_qos` and `sub_qos`, so would it not make more sense to copy either of those instead?

There is also the (remote) possibility that this call to `get_readwrite_qos` fails: if some other thread calls `rmw_destroy_client` between line 5117 and this line, it might. Not calling it avoids having to consider the possibility and whether it should be taken seriously.

Reply:
Right, I'm copying the QoS settings now (last commit 1acdd5c)